api_app/analyzers_manager/observable_analyzers/onionscan.py

```diff
@@ -0,0 +1,107 @@
+    """
+
+    onionscan_binary: str = "/opt/deploy/bundled/onionscan"
+    _verbose: bool = True
+    _tor_proxy_address: str = None
+
+    def set_params(self, params):
+        self._verbose = params.get("verbose", True)
+        self._tor_proxy_address = params.get("torProxyAddress", None)
+
+    def run(self):
+        """
+        Run Onionscan against target onion domain
+        """
+        # Check for onionscan binary in bundled binaries.
+        if not which(self.onionscan_binary):
+            raise AnalyzerRunException("onionscan is not installed!")
+        # Generate the subprocess command args
+        onionscan_cmd = [
+            self.onionscan_binary,
+        ]
+        if self._verbose:
+            onionscan_cmd.append("-verbose")
+        if self._tor_proxy_address:
+            onionscan_cmd.extend(["-torProxyAddress", self._tor_proxy_address])
+        # Open a pipe to onionscan process
+        onionscan_cmd.extend(["-jsonReport", self.observable_name])
+        if not self.check_tor():
+            raise AnalyzerRunException("tor not yet ready.")
+        try:
+            process = subprocess.Popen(
+                onionscan_cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            # read all std communication
+            outs, errs = [std.decode("utf-8") for std in process.communicate()]
+            if errs:
+                raise AnalyzerRunException(
+                    f"onionscan params: {''.join(onionscan_cmd)}."
+                    + f" onionscan error: {str(errs)}."
+                )
+            # load stdout json and return to user
+            logger.info("onionscan output: \n%s", outs)
+            onionscan_json_report = json.loads(outs)
+            # return report to user
+            return onionscan_json_report
+        except subprocess.SubprocessError as exc:
+            raise AnalyzerRunException(
+                f"error spwaning onionscan process. error: {str(exc)}"
+            )
+        except json.decoder.JSONDecodeError as exc:
+            raise AnalyzerRunException(
+                f"error reading onionscan output json. error: {str(exc)}"
+            )
+        except SoftTimeLimitExceeded as exc:
+            # handle celery timeout
+            self._handle_exception(exc)
+            if process:
+                process.kill()
+
+    def check_tor(self):
+        """
+        check if tor bootstrap is complete
+        """
+        try:
+            url = "https://check.torproject.org/api/ip"
+            tor_proxy = {
+                "http": "socks5h://127.0.0.1:9050",
+                "https": "socks5h://127.0.0.1:9050",
+            }
+            r = requests.get(url, proxies=tor_proxy)
+            IsTor = json.loads(r.text)["IsTor"]
+            return IsTor
+        except Exception as e:
+            logger.info("onionscan:check_tor error: \n%s", str(e))
```

Comment:
use the most recent f-format, not the old %s please

Reply:
Ok, Sir.